# Finding Concurrency Problems with the Data Race Detector

- It's easy to accidentally reference a variable from two different goroutines without acquring a lock.
- That is a **data race**.
- To help find these sorts of bugs, Go includes a **race checker**.
- It isn't guaranteed to find every single data race in our code, but if it finds one, we should put proper locks around what it finds.



`race/race.go`

```go
package race

import "sync"

func getCounter() int{
	var counter int 
	var wg sync.WaitGroup
	wg.Add(5)
	for i := 0; i < 5; i++{
		go func(){
			counter++
		}()
		wg.Done()
	}
	wg.Wait()
	return counter
}
```

`race/race_test.go`

```go
package race

import "testing"

func TestGetCounter(t *testing.T){
	counter := getCounter()
	if counter != 5000{
		t.Error("unexpected counter: ", counter)
	}
}
```

`Output`

- In our program, multiple goroutines are trying to update `counter` simultaneously, and some of their updates are lost.

```sh
race âž¤ go test
--- FAIL: TestGetCounter (0.00s)
    race_test.go:8: unexpected counter:  2
FAIL
exit status 1
FAIL    practice/race   0.001s
```

```sh
==================
==================
WARNING: DATA RACE
Read at 0x00c000014278 by goroutine 11:
  practice/race.getCounter.func1()
      /home/chan/Go_Programming/practice/race/race.go:11 +0x2e

Previous write at 0x00c000014278 by goroutine 10:
  practice/race.getCounter.func1()
      /home/chan/Go_Programming/practice/race/race.go:11 +0x44
```

